pkg/authorization/api/validation/validation.go

```diff
@@ -8,48 +8,126 @@ import (
 	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
 )
 
-// ValidatePolicyBinding tests required fields for a PolicyBinding.
-func ValidatePolicyBinding(policyBinding *authorizationapi.PolicyBinding) fielderrors.ValidationErrorList {
+func ValidatePolicyName(name string, prefix bool) (bool, string) {
+	if name != authorizationapi.PolicyName {
+		return false, "name must be " + authorizationapi.PolicyName
+	}
+
+	return true, ""
+}
+
+func ValidatePolicy(policy *authorizationapi.Policy) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
+	allErrs = append(allErrs, validation.ValidateObjectMeta(&policy.ObjectMeta, true, ValidatePolicyName).Prefix("metadata")...)
+
+	for roleKey, role := range policy.Roles {
+		if roleKey != role.Name {
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("roles."+roleKey+".name", policy, "must be "+roleKey))
+		}
+
+		allErrs = append(allErrs, ValidateRole(&role).Prefix("roles."+roleKey)...)
+	}
+
+	return allErrs
+}
 
-	if len(policyBinding.Name) == 0 {
-		allErrs = append(allErrs, fielderrors.NewFieldRequired("name"))
+func ValidatePolicyUpdate(oldPolicy *authorizationapi.Policy, policy *authorizationapi.Policy) fielderrors.ValidationErrorList {
+	allErrs := ValidatePolicy(policy)
+	allErrs = append(allErrs, validation.ValidateObjectMetaUpdate(&oldPolicy.ObjectMeta, &policy.ObjectMeta).Prefix("metadata")...)
```

Comment:
ValidateObjectMetaUpdate requires resource version... what are you deriving the resource version of a role from?